src/native-state-qnx.cpp

```diff
@@ -0,0 +1,218 @@
+
+    destroy_window();
+
+    rc = screen_create_window(&screen_win_, screen_ctx_);
+    if (rc) {
+        Log::error("screen_create_window failed, rc=%d", rc);
+        return false;
+	}
+
+    rc = screen_set_window_property_cv(screen_win_,
+            SCREEN_PROPERTY_ID_STRING, strlen(win_name), win_name);
+    if (rc) {
+        Log::error("screen_set_window_property_cv(SCREEN_PROPERTY_ID_STRING) failed, rc=%d", rc);
+        return false;
+    }
+
+    int size[2] = { properties.width, properties.height };
+    if (!properties.fullscreen && size[0] > 0 && size[1] > 0) {
+        rc = screen_set_window_property_iv(screen_win_, SCREEN_PROPERTY_SIZE, size);
+        if (rc) {
+            Log::error("screen_set_window_property_iv(SCREEN_PROPERTY_SIZE) failed. rc=%d, size={%d,%d}",
+                        rc, size[0], size[1]);
+            return false;
+        }
+        win_properties_.fullscreen = false;
+    } else {
+        win_properties_.fullscreen = true;
+    }
+
+    int usage = SCREEN_USAGE_OPENGL_ES2;
+    rc = screen_set_window_property_iv(screen_win_, SCREEN_PROPERTY_USAGE, &usage);
+    if (rc) {
+        Log::error("screen_set_window_property_iv(SCREEN_PROPERTY_USAGE) failed. rc=%d", rc);
+        return false;
+    }
+
+    int format = properties.visual_id;
+    switch(format) {
+    case SCREEN_FORMAT_RGBA8888:
+    case SCREEN_FORMAT_RGBX8888:
+    case SCREEN_FORMAT_RGB565:
+        win_properties_.visual_id = format;
+        break;
+    default:
+        Log::error("unsupported native visual_id. %d", properties.visual_id);
+        return false;
+    }
+
+    rc = screen_set_window_property_iv(screen_win_, SCREEN_PROPERTY_FORMAT, &format);
+    if (rc) {
+        Log::error("screen_set_window_property_iv(SCREEN_PROPERTY_FORMAT), rc=%d", rc);
+        return false;
+    }
+
+    Log::debug("Creating Window buffers W: %d H: %d FMT: %d\n",
+                properties.width, properties.height, format);
+
+    int nbuffers = 2;
+    rc = screen_create_window_buffers(screen_win_, nbuffers);
+    if (rc) {
+        Log::error("screen_create_window_buffers failed, rc=%d", rc);
+        return false;
+    }
+
+    if (win_properties_.fullscreen) {
+        rc = screen_get_window_property_iv(screen_win_, SCREEN_PROPERTY_SIZE, size);
+        if (rc) {
+            Log::error("screen_get_window_property_iv(SCREEN_PROPERTY_SIZE)  failed, rc=%d", rc);
+            return false;
+        }
+    }
+    win_properties_.width  = size[0];
+    win_properties_.height = size[1];
+
+    rc = screen_set_window_property_iv(screen_win_, SCREEN_PROPERTY_VISIBLE, &win_visible_);
+    if (rc) {
+        Log::error("screen_set_window_property_iv(SCREEN_PROPERTY_VISIBLE)  failed, rc=%d", rc);
+    }
+
+    return true;
+}
+
+void*
+NativeStateQnx::window(WindowProperties& properties)
+{
+    properties = win_properties_;
+    return (void *)screen_win_;
+}
+
+void
+NativeStateQnx::visible(bool visible)
+{
+    win_visible_ = visible ? 1 : 0;
```

Comment:
The core glmark core code expects that calling native_state.visible(true) will show the window without any additional calls (i.e. you can't expect that the core will call create_window() after visible(true)). The correct way to set this up would be for win_visible_  to start as false and then visible() to call screen_get_window_property_iv(..., SCREEN_PROPERTY_VISIBLE, ...) to show/hide the window.

Note that the current code happens to work because it makes the window visible from the start, but it's not strictly correct in terms of NativeState API.